File: frontend/src/lib/game/optimizations.ts

export class SpatialHash {
  private cellSize: number
  private grid: Map<string, Map<string, [x: number, y: number]>>
  private playerCells: Map<string, string> 
  
  constructor(cellSize: number = 400) {
    this.cellSize = cellSize
    this.grid = new Map()
    this.playerCells = new Map()
  }

  private hash(x: number, y: number): string {
    const cellX = Math.floor(x / this.cellSize)
    const cellY = Math.floor(y / this.cellSize)
    return `${cellX},${cellY}`
  }

  update(username: string, x: number, y: number ) {
    const newKey = this.hash(x, y)
    const oldKey = this.playerCells.get(username)
    
    if (oldKey && oldKey !== newKey) {
      const oldCell = this.grid.get(oldKey)
      if (oldCell) {
        oldCell.delete(username)
        if (oldCell.size === 0) {
          this.grid.delete(oldKey)
        }
      }
    }
    if (!this.grid.has(newKey)) {
      this.grid.set(newKey, new Map())
    }
    
    this.grid.get(newKey)!.set(username, [x, y])
    this.playerCells.set(username, newKey)
  }

  remove(username: string) {
    const key = this.playerCells.get(username)
    
    if (key) {
      const cell = this.grid.get(key)
      if (cell) {
        cell.delete(username)
        if (cell.size === 0) {
          this.grid.delete(key)
        }
      }
      this.playerCells.delete(username)
    }
  }

  getCell(x: number, y: number): Map<string, [x: number, y: number]> {
    const key = this.hash(x, y)
    return this.grid.get(key) || new Map()
  }

  clear() {
    this.grid.clear()
    this.playerCells.clear()
  }
}
export const FastProjectileCheck = new SpatialHash(400)

File: backend/optimizations.go
package main

import "sync"

type Vec2 struct {
	X float32
	Y float32
}

type SpatialHash struct {
	cellSize   float32
	invCell    float32
	grid       map[int64]map[string]Vec2
	playerCell map[string]int64
	mu         sync.RWMutex
}

func NewSpatialHash(cellSize float32) *SpatialHash {
	return &SpatialHash{
		cellSize:   cellSize,
		invCell:    1.0 / cellSize,
		grid:       make(map[int64]map[string]Vec2, 1024),
		playerCell: make(map[string]int64, 1024),
	}
}

func (s *SpatialHash) hash(x, y float32) int64 {
	cx := int32(x * s.invCell)
	cy := int32(y * s.invCell)
	return cellKey(cx, cy)
}

func (s *SpatialHash) Update(id string, x, y float32) {
	key := s.hash(x, y)

	if oldKey, ok := s.playerCell[id]; ok && oldKey != key {
		if cell := s.grid[oldKey]; cell != nil {
			delete(cell, id)
			if len(cell) == 0 {
				delete(s.grid, oldKey)
			}
		}
	}

	cell := s.grid[key]
	if cell == nil {
		cell = make(map[string]Vec2, 8)
		s.grid[key] = cell
	}

	cell[id] = Vec2{x, y}
	s.playerCell[id] = key
}

func (s *SpatialHash) Remove(id string) {
	key, ok := s.playerCell[id]
	if !ok {
		return
	}
	if cell := s.grid[key]; cell != nil {
		delete(cell, id)
		if len(cell) == 0 {
			delete(s.grid, key)
		}
	}

	delete(s.playerCell, id)
}

func (s *SpatialHash) GetCell(x, y float32) map[string]Vec2 {
	return s.grid[s.hash(x, y)]
}

func cellKey(cx, cy int32) int64 {
	return int64(cx)<<32 | int64(uint32(cy))
}
